src/StreamrClient.js

```diff
@@ -702,12 +797,31 @@ export default class StreamrClient extends EventEmitter {
         }
     }
 
+    async publishStreamMessage(streamMessage) {
+        const sessionToken = await this.session.getSessionToken()
+        return this._requestPublish(streamMessage, sessionToken)
+    }
+
     _requestPublish(streamMessage, sessionToken) {
         const request = ControlLayer.PublishRequest.create(streamMessage, sessionToken)
         debug('_requestPublish: %o', request)
         return this.connection.send(request)
     }
 
+    // each element of the array "groupKeys" is an object with 2 fields: "groupKey" and "start"
+    _setGroupKeys(streamId, publisherId, groupKeys) {
+        if (!this.options.subscriberGroupKeys[streamId]) {
+            this.options.subscriberGroupKeys[streamId] = {}
+        }
+        const last = groupKeys[groupKeys.length - 1]
+        const current = this.options.subscriberGroupKeys[streamId][publisherId]
+        if (!current || last.start > current.start) {
+            this.options.subscriberGroupKeys[streamId][publisherId] = last
+        }
+        // TODO: fix this hack in other PR
+        this.subscribedStreamPartitions[streamId + '0'].setSubscriptionsGroupKeys(publisherId, groupKeys.map((obj) => obj.groupKey))
```

Comment:
does this hack still need a fix?

Reply:
Yes this is not related to key-exchange and should be done once this PR lands.